File: converted/source2/source/zombie/fileSystem/TexturePackDevice.java
package zombie.fileSystem;

import java.io.BufferedInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import zombie.ZomboidFileSystem;
import zombie.core.textures.TexturePackPage;
import zombie.core.textures.TexturePackPage.SubTextureInfo;
import zombie.fileSystem.FileSystem.TexturePackTextures;
import zombie.fileSystem.TexturePackDevice.Page;
import zombie.fileSystem.TexturePackDevice.PositionInputStream;
import zombie.fileSystem.TexturePackDevice.SubTexture;
import zombie.fileSystem.TexturePackDevice.TexturePackInputStream;

public final class TexturePackDevice implements IFileDevice {
   static final int VERSION1 = 1;
   static final int VERSION_LATEST = 1;
   String m_name;
   String m_filename;
   int m_version = -1;
   final ArrayList<Page> m_pages = new ArrayList<>();
   final HashMap<String, Page> m_pagemap = new HashMap<>();
   final HashMap<String, SubTexture> m_submap = new HashMap<>();
   int m_textureFlags;

   private static long skipInput(InputStream var0, long var1) throws IOException {
      long var3 = 0L;

      while (var3 < var1) {
         long var5 = var0.skip(var1 - var3);
         if (var5 > 0L) {
            var3 += var5;
         }
      }

      return var3;
   }

   public TexturePackDevice(String var1, int var2) {
      this.m_name = var1;
      this.m_filename = ZomboidFileSystem.instance.getString("media/texturepacks/" + var1 + ".pack");
      this.m_textureFlags = var2;
   }

   public IFile createFile(IFile var1) {
      return null;
   }

   public void destroyFile(IFile var1) {
   }

   public InputStream createStream(String var1, InputStream var2) throws IOException {
      this.initMetaData();
      return new TexturePackInputStream(var1, this);
   }

   public void destroyStream(InputStream var1) {
      if (var1 instanceof TexturePackInputStream) {
      }
   }

   public String name() {
      return this.m_name;
   }

   public void getSubTextureInfo(TexturePackTextures var1) throws IOException {
      this.initMetaData();

      for (SubTexture var3 : this.m_submap.values()) {
         zombie.fileSystem.FileSystem.SubTexture var4 = new zombie.fileSystem.FileSystem.SubTexture(this.name(), var3.m_page.m_name, var3.m_info);
         var1.put(var3.m_info.name, var4);
      }
   }

   private void initMetaData() throws IOException {
      if (this.m_pages.isEmpty()) {
         try (
            FileInputStream var1 = new FileInputStream(this.m_filename);
            BufferedInputStream var2 = new BufferedInputStream(var1);
         ) {
            PositionInputStream var3 = new PositionInputStream(this, var2);

            try {
               var3.mark(4);
               int var4 = var3.read();
               int var5 = var3.read();
               int var6 = var3.read();
               int var7 = var3.read();
               if (var4 == 80 && var5 == 90 && var6 == 80 && var7 == 75) {
                  this.m_version = TexturePackPage.readInt(var3);
                  if (this.m_version < 1 || this.m_version > 1) {
                     throw new IOException("invalid .pack file version " + this.m_version);
                  }
               } else {
                  var3.reset();
                  this.m_version = 0;
               }

               int var8 = TexturePackPage.readInt(var3);

               for (int var9 = 0; var9 < var8; var9++) {
                  Page var10 = this.readPage(var3);
                  this.m_pages.add(var10);
                  this.m_pagemap.put(var10.m_name, var10);

                  for (SubTextureInfo var12 : var10.m_sub) {
                     this.m_submap.put(var12.name, new SubTexture(var10, var12));
                  }
               }
            } catch (Throwable var16) {
               try {
                  var3.close();
               } catch (Throwable var15) {
                  var16.addSuppressed(var15);
               }

               throw var16;
            }

            var3.close();
         }
      }
   }

   private Page readPage(PositionInputStream var1) throws IOException {
      Page var2 = new Page();
      String var3 = TexturePackPage.ReadString(var1);
      int var4 = TexturePackPage.readInt(var1);
      boolean var5 = TexturePackPage.readInt(var1) != 0;
      var2.m_name = var3;
      var2.m_has_alpha = var5;

      for (int var6 = 0; var6 < var4; var6++) {
         String var7 = TexturePackPage.ReadString(var1);
         int var8 = TexturePackPage.readInt(var1);
         int var9 = TexturePackPage.readInt(var1);
         int var10 = TexturePackPage.readInt(var1);
         int var11 = TexturePackPage.readInt(var1);
         int var12 = TexturePackPage.readInt(var1);
         int var13 = TexturePackPage.readInt(var1);
         int var14 = TexturePackPage.readInt(var1);
         int var15 = TexturePackPage.readInt(var1);
         var2.m_sub.add(new SubTextureInfo(var8, var9, var10, var11, var12, var13, var14, var15, var7));
      }

      var2.m_png_start = var1.getPosition();
      if (this.m_version == 0) {
         int var16 = 0;

         do {
            var16 = TexturePackPage.readIntByte(var1);
         } while (var16 != -559038737);
      } else {
         int var18 = TexturePackPage.readInt(var1);
         skipInput(var1, var18);
      }

      return var2;
   }

   public boolean isAlpha(String var1) {
      Page var2 = this.m_pagemap.get(var1);
      return var2.m_has_alpha;
   }

   public int getTextureFlags() {
      return this.m_textureFlags;
   }
}

File: ProjectZomboidSwitch/src/zombie/fileSystem/TexturePackDevice.cpp
#include <string>
#include "zombie/fileSystem/TexturePackDevice.h"

namespace zombie {
namespace fileSystem {

long TexturePackDevice::skipInput(InputStream var0, long var1) {
    long total = 0;
    while (total < var1) {
        long skipped = var0->skip(var1 - total);
        if (skipped > 0) {
            total += skipped;
        } else {
            break;
        }
    }
    return total;
}
}

TexturePackDevice::TexturePackDevice(const std::string& var1, int var2)
    : m_name(var1), m_filename("media/texturepacks/" + var1 + ".pack"), m_textureFlags(var2), m_version(-1)
{
    m_isLoaded = false;
    m_packSize = 0;
    m_packOffset = 0;
    m_packStream = nullptr;
    m_pages.clear();
    m_submap.clear();
    m_pagemap.clear();
    // Java/header fields
    // m_version is already set to -1 in initializer
    // m_textureFlags is set from var2
    // m_name and m_filename are set from var1
    // Add any additional fields as needed
}
}

IFile TexturePackDevice::createFile(IFile var1) {
    // TexturePackDevice is read-only; cannot create files in a texture pack
    // Optionally, throw or log an error here
    return nullptr;
}
}

void TexturePackDevice::destroyFile(IFile var1) {
    // No-op
}
}

InputStream TexturePackDevice::createStream(const std::string& var1, InputStream var2) {
    initMetaData();
    // TexturePackInputStream should be constructed with the texture name and this device
    return std::make_shared<TexturePackInputStream>(var1, shared_from_this());
}
}

void TexturePackDevice::destroyStream(InputStream var1) {
    // If var1 is a TexturePackInputStream, reset the pointer to release resources
    auto texStream = std::dynamic_pointer_cast<TexturePackInputStream>(var1);
    if (texStream) {
        texStream.reset();
    }
}
}

std::string TexturePackDevice::name() {
    return m_name;
}
}

void TexturePackDevice::getSubTextureInfo(TexturePackTextures var1) {
    initMetaData();
    for (const auto& pair : m_submap) {
        const auto& key = pair.first;
        const auto& sub = pair.second;
        // Construct a FileSystem::SubTexture (pack name, page name, subtexture info)
        FileSystem::SubTexture fsSubTex(name(), sub.m_page.m_name, sub.m_info);
        (*var1)[key] = fsSubTex;
    }
}
}

void TexturePackDevice::initMetaData() {
    if (!m_pages.empty()) {
        return;
    }

    // Open the file
    std::ifstream file(m_filename, std::ios::binary);
    if (!file.is_open()) {
        // Could not open file, handle error as needed
        return;
    }

    // Read first 4 bytes
    char magic[4];
    file.read(magic, 4);
    if (!file) {
        file.close();
        return;
    }

    if (magic[0] == 'P' && magic[1] == 'Z' && magic[2] == 'P' && magic[3] == 'K') {
        // Read version (int32)
        int32_t version = 0;
        file.read(reinterpret_cast<char*>(&version), sizeof(version));
        m_version = version;
        if (m_version < 1 || m_version > 1) {
            file.close();
            return;
        }
    } else {
        // Not a PZPK file, reset and set version 0
        file.seekg(0);
        m_version = 0;
    }

    // Read number of pages (int32)
    int32_t numPages = 0;
    file.read(reinterpret_cast<char*>(&numPages), sizeof(numPages));
    if (!file) {
        file.close();
        return;
    }


    // Wrap the file stream in a PositionInputStream (if available)
    // For now, assume file is compatible with PositionInputStream constructor
    auto posStream = std::make_shared<PositionInputStream>(shared_from_this(), file);
    for (int i = 0; i < numPages; ++i) {
        Page page = readPage(posStream);
        m_pages.push_back(page);
        m_pagemap[page->m_name] = page;
        for (const auto& sub : page->m_sub) {
            m_submap[sub->name] = SubTexture(page, *sub);
        }
    }

    file.close();
}
}

Page TexturePackDevice::readPage(PositionInputStream var1) {
    // Create a new Page object
    auto page = std::make_shared<Page>();

    // Read page name
    page->m_name = TexturePackPage::ReadString(var1);
    // Read number of subtextures
    int numSubs = TexturePackPage::readInt(var1);
    // Read has_alpha
    page->m_has_alpha = TexturePackPage::readInt(var1) != 0;

    // Read subtextures
    for (int i = 0; i < numSubs; ++i) {
        std::string subName = TexturePackPage::ReadString(var1);
        int x = TexturePackPage::readInt(var1);
        int y = TexturePackPage::readInt(var1);
        int w = TexturePackPage::readInt(var1);
        int h = TexturePackPage::readInt(var1);
        int ox = TexturePackPage::readInt(var1);
        int oy = TexturePackPage::readInt(var1);
        int fx = TexturePackPage::readInt(var1);
        int fy = TexturePackPage::readInt(var1);
        auto subInfo = std::make_shared<SubTextureInfo>(x, y, w, h, ox, oy, fx, fy, subName);
        page->m_sub.push_back(subInfo);
    }

    // Set PNG start position
    page->m_png_start = var1->getPosition();

    if (m_version == 0) {
        int marker = 0;
        do {
            marker = TexturePackPage::readIntByte(var1);
        } while (marker != -559038737);
    } else {
        int skipLen = TexturePackPage::readInt(var1);
        skipInput(var1, skipLen);
    }

    return page;
}
}

bool TexturePackDevice::isAlpha(const std::string& var1) {
    auto it = m_pagemap.find(var1);
    if (it != m_pagemap.end() && it->second) {
        return it->second->m_has_alpha;
    }
    return false;
}
}

int TexturePackDevice::getTextureFlags() {
    return m_textureFlags;
}
}

} // namespace fileSystem
} // namespace zombie
